### PR TITLE
Unit test for MediaScannerService

### PR DESCRIPTION
**Pull request by biconou**
This PR adds a minimal media library for unit test only and a unit test case for the MediaScannerService.

The media library is located into the src/test/recources/MEDIAS directory. 
It is composed of little audio files, covers images organized in albums.
The folders are named with the _DIR_ prefix and certain ID3 tags with the _ID3_ prefix so it is easy to understand the maning source of objects in the database. 

The MediaScannerService test case scans the library, reports some informations about the it's contains and tests some assertions (to be continued).

The default libresonic_home path used for testing has been refactored and changed. Indeed I don't think it is a good idea to keep /tmp/libresonic as home directory. Now it will be set to the target/tmp/libresonic directory inside the libresonic-main project hierarchy (accoding to maven).

### CODE REVIEW

**Comment by alexfornuto** [comment]
I'm not a dev, but I'm interested in helping. Assuming I build using this branch, what can I do to test the functionality?

**Comment by biconou** [comment]
You'll only have to run maven tests. Or, using IntelliJ, right click on the MediaScannerServiceTestCase class and choose run (or debug). the output of the test will appear in the console window.
Note that's it is just the minimal version of this test. I'll add some assertions/reports in a near future.
Don't forget to suggest some other informations/results you'd expect from this test.

**Comment by alexfornuto** [comment]
Sounds like it's outside of my scope of knowledge. 😢

**Comment by biconou** [comment]
Do you build using the command : mvn package ?

**Comment by alexfornuto** [comment]
I did...

**Comment by biconou** [comment]
Running mvn package, you'll see (among other things) the following lines : 

> Running org.libresonic.player.service.MediaScannerServiceTestCase
> Delete libresonic home (ie. /develop/libresonic/libresonic-main/target/tmp/libresonic).
> --- Report of records count per table ---
> PODCAST_CHANNEL : 0
> USER_SETTINGS : 0
> SHARE : 0
> TRANSCODING : 19
> ROLE : 11
> MEDIA_FILE : 20
> PLAY_QUEUE : 0
> ARTIST : 5
> MUSIC_FOLDER_USER : 1
> MUSIC_FILE_INFO : 0
> PLAYER_TRANSCODING : 0
> STARRED_ALBUM : 0
> BOOKMARK : 0
> SHARE_FILE : 0
> MUSIC_FOLDER : 1
> INTERNET_RADIO : 0
> GENRE : 5
> USER : 1
> PLAYLIST_FILE : 0
> PLAYER_TRANSCODING2 : 0
> PODCAST_EPISODE : 0
> PLAYER : 0
> CUSTOM_AVATAR : 0
> PLAY_QUEUE_FILE : 0
> PLAYLIST_USER : 0
> STARRED_ARTIST : 0
> VERSION : 25
> USER_RATING : 0
> ALBUM : 5
> USER_ROLE : 11
> TRANSCODING2 : 3
> STARRED_MEDIA_FILE : 0
> PLAYLIST : 0
> --- *********************** ---
> --- List of all artists ---
> artistName#albumCount
> _DIR_ Ravel#1
> _ID3_ALBUMARTIST_ Chrome Hoof#1
> _ID3_ALBUMARTIST_ Céline Frisch: Café Zimmermann#1
> _ID3_ALBUMARTIST_ Sarah Walker/Nash Ensemble#1
> _ID3_ALBUMARTIST_ Sixteen Horsepower#1
> --- *********************** ---
> --- List of all albums ---
> name#artist
> Complete Piano Works#_DIR_ Ravel
> _ID3_ALBUM_ Chrome Hoof#_ID3_ALBUMARTIST_ Chrome Hoof
> _ID3_ALBUM_ Bach: Goldberg Variations, Canons [Disc 1]#_ID3_ALBUMARTIST_ Céline Frisch: Café Zimmermann
> _ID3_ALBUM_ Ravel - Chamber Music With Voice#_ID3_ALBUMARTIST_ Sarah Walker/Nash Ensemble
> _ID3_ALBUM_ Sackcloth 'n' Ashes#_ID3_ALBUMARTIST_ Sixteen Horsepower
> --- *********************** ---
> 07/09/16 23:37:09 ==============================================================
> 
> -- Timers ----------------------------------------------------------------------
> org.libresonic.player.service.MediaScannerServiceTestCase.Timer.global
>              count = 1
>          mean rate = 0,92 calls/second
>      1-minute rate = 0,00 calls/second
>      5-minute rate = 0,00 calls/second
>     15-minute rate = 0,00 calls/second
>                min = 1000,53 milliseconds
>                max = 1000,53 milliseconds
>               mean = 1000,53 milliseconds
>             stddev = 0,00 milliseconds
>             median = 1000,53 milliseconds
>               75% <= 1000,53 milliseconds
>               95% <= 1000,53 milliseconds
>               98% <= 1000,53 milliseconds
>               99% <= 1000,53 milliseconds
>             99.9% <= 1000,53 milliseconds
> 
> 
> EndTests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 3.28 sec
> 
> 

Fisrt, you see the number of records in each table after scanning the tiny media library. 
Then a list of all discovered artits and albums. 
A line like 

> _ID3_ALBUMARTIST_ Chrome Hoof#1

teaches us that the artist "Chrome Hoof" has been discovered from a ALBUMARTIST ID3 tag (mata-data in mp3 file), and that there are at least 1 album for that artist.


Finally, some statistics of the duration of scanning process (in my case, around one second).

**Comment by alexfornuto** [comment]
I'll give it a shot tonight.

File: libresonic-main/pom.xml [hunk 1]
```diff
@@ -26,6 +30,15 @@
...
         </dependency>
 
+        <!-- Metrics -->
+        <dependency>
+            <groupId>io.dropwizard.metrics</groupId>
+            <artifactId>metrics-core</artifactId>
+            <version>${metrics.version}</version>
+            <scope>test</scope>
+        </dependency>
+
```
Comment (muff1nman) [inline libresonic-main/pom.xml]: This seems unrelated to the tests added here. Do we really need this? Can this go in a separate PR?

File: libresonic-main/src/test/java/org/libresonic/player/TestCaseUtils.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+  private static final String LIBRESONIC_HOME_FOR_TEST = "/tmp/libresonic";
+
+  public static String libresonicHomePathForTest() {
+
+    String targetDir = TestCaseUtils.class.getResource("/").toString().replaceFirst("/target.*","/target");
+    if (targetDir.startsWith("file:")) {
+      targetDir = targetDir.replace("file:","");
+    }
+    return targetDir + LIBRESONIC_HOME_FOR_TEST;
+  }
```
Comment (muff1nman) [inline libresonic-main/src/test/java/org/libresonic/player/TestCaseUtils.java]: So previously we were using /tmp/libresonic for tmp files, and this would change that to instead be in the target directory. Not terrible, but would it not be better to use a java tmp directory. That way we don't muck with maven, but still allow for more portability.

File: libresonic-main/src/test/java/org/libresonic/player/TestCaseUtils.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+                    "from information_schema.system_tables " +
+                    "where table_name not like 'SYSTEM%'"
+            , String.class);
+    Map<String, Integer> nbRecords =
+            tableNames.stream()
+                    .collect(Collectors.toMap(table -> table, table -> recordsInTable(table,daoHelper)));
+
+    return nbRecords;
+  }
+
```
Comment (muff1nman) [inline libresonic-main/src/test/java/org/libresonic/player/TestCaseUtils.java]: This method (recordsInAllTables) doesn't seem to be used by the tests. Can we remove it?

File: libresonic-main/src/test/java/org/libresonic/player/TestCaseUtils.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+   * Counts records in a table.
+   *
+   * @param tableName
+   * @param daoHelper
+   * @return
+     */
+  public static Integer recordsInTable(String tableName, DaoHelper daoHelper) {
+    return daoHelper.getJdbcTemplate().queryForInt("select count(1) from " + tableName);
+  }
+
```
Comment (muff1nman) [inline libresonic-main/src/test/java/org/libresonic/player/TestCaseUtils.java]: Same as above.

File: libresonic-main/src/test/java/org/libresonic/player/TestCaseUtils.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+      try {
+        Thread.sleep(1000);
+      } catch (InterruptedException e) {
+        e.printStackTrace();
+      }
+    }
+
+  }
+
+}
```
Comment (muff1nman) [inline libresonic-main/src/test/java/org/libresonic/player/TestCaseUtils.java]: In fact, most of this class doesn't seem to be used anywhere...

Comment (muff1nman) [inline libresonic-main/src/test/java/org/libresonic/player/TestCaseUtils.java]: Scratch this comment. Intellij was messing with me.

**Comment by muff1nman** [comment]
```
$ du -d 1 -h
2.8M	./_DIR_ Céline Frisch- Café Zimmermann - Bach- Goldberg Variations, Canons [Disc 1]
2.8M	./_DIR_ Sixteen Horsepower
2.2M	./_DIR_ Ravel
7.7M	.
```
Any way we can shrink down the size of those media files? Maybe chop em to 10 seconds or something. I understand the need to have something for the tests, but hate to have our repository size grow too much if we can be smart about it.

File: libresonic-main/src/test/java/org/libresonic/player/TestCaseUtils.java [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+  /**
+   * Scans the music library   * @param mediaScannerService
+   */
+  public static void execScan(MediaScannerService mediaScannerService) {
+    mediaScannerService.scanLibrary();
+
+    while (mediaScannerService.isScanning()) {
+      try {
+        Thread.sleep(1000);
```
Comment (muff1nman) [inline libresonic-main/src/test/java/org/libresonic/player/TestCaseUtils.java]: Hmm, unfortunate that we have to busy wait but I see why you had to do this considering the implementation behind MediaScannerService. I dont think we can improve on this for now, but maybe in the future we can.

File: libresonic-main/src/test/java/org/libresonic/player/service/MediaScannerServiceTestCase.java [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+    System.out.println("--- List of all albums ---");
+    System.out.println("name#artist");
+    List<Album> allAlbums = albumDao.getAlphabetialAlbums(0,0,true,musicFolderDao.getAllMusicFolders());
+    allAlbums.forEach(album -> System.out.println(album.getName()+"#"+album.getArtist()));
+    Assert.assertEquals(5,allAlbums.size());
+    System.out.println("--- *********************** ---");
+
+
+
+    List<MediaFile> listeSongs = mediaFileDao.getSongsByGenre("Baroque Instrumental",0,0,musicFolderDao.getAllMusicFolders());
```
Comment (muff1nman) [inline libresonic-main/src/test/java/org/libresonic/player/service/MediaScannerServiceTestCase.java]: Doesn't seem to be used.

File: libresonic-main/src/test/java/org/libresonic/player/service/MediaScannerServiceTestCase.java [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+            .build();
+    reporter.report();
+
+    System.out.print("End");
+  }
+
+  /**
+   * Starts a Metrics Console and JMX reporter.
+   */
+  private void startMetricsReport() {
```
Comment (muff1nman) [inline libresonic-main/src/test/java/org/libresonic/player/service/MediaScannerServiceTestCase.java]: Doesnt seem to be used, do we need it?

**Comment by biconou** [comment]
@muff1nman Thanks for this review. I'll correct the hole PR.

File: libresonic-main/pom.xml (hunk 1 above)
Comment (biconou) [inline libresonic-main/pom.xml]: Metrics is a very, very usefull framework to monitor performance issues. I really suggest to add it to Libresonic.


In the context of this unit test it reports this.

-- Timers ----------------------------------------------------------------------
org.libresonic.player.service.MediaScannerServiceTestCase.Timer.global
count = 1
mean rate = 0,92 calls/second
1-minute rate = 0,00 calls/second
5-minute rate = 0,00 calls/second
15-minute rate = 0,00 calls/second
min = 1000,53 milliseconds
max = 1000,53 milliseconds
mean = 1000,53 milliseconds
stddev = 0,00 milliseconds
median = 1000,53 milliseconds
75% <= 1000,53 milliseconds
95% <= 1000,53 milliseconds
98% <= 1000,53 milliseconds
99% <= 1000,53 milliseconds
99.9% <= 1000,53 milliseconds

File: libresonic-main/src/test/java/org/libresonic/player/TestCaseUtils.java (hunk 2 above)
Comment (biconou) [inline libresonic-main/src/test/java/org/libresonic/player/TestCaseUtils.java]: I'll consider using a java tmp dir.

File: libresonic-main/src/test/java/org/libresonic/player/TestCaseUtils.java (hunk 3 above)
Comment (biconou) [inline libresonic-main/src/test/java/org/libresonic/player/TestCaseUtils.java]: I must inspect this point.

File: libresonic-main/src/test/java/org/libresonic/player/TestCaseUtils.java (hunk 2 above)
Comment (biconou) [inline libresonic-main/src/test/java/org/libresonic/player/TestCaseUtils.java]: Fixed

File: libresonic-main/src/test/java/org/libresonic/player/TestCaseUtils.java (hunk 3 above)
Comment (biconou) [inline libresonic-main/src/test/java/org/libresonic/player/TestCaseUtils.java]: Ok, it is actually used and usefull I think as it reports the exact count of records per table after mediascan.

File: libresonic-main/src/test/java/org/libresonic/player/TestCaseUtils.java (hunk 4 above)
Comment (biconou) [inline libresonic-main/src/test/java/org/libresonic/player/TestCaseUtils.java]: ok

File: libresonic-main/src/test/java/org/libresonic/player/TestCaseUtils.java (hunk 5 above)
Comment (biconou) [inline libresonic-main/src/test/java/org/libresonic/player/TestCaseUtils.java]: ok

File: libresonic-main/src/test/java/org/libresonic/player/service/MediaScannerServiceTestCase.java (hunk 8 above)
Comment (biconou) [inline libresonic-main/src/test/java/org/libresonic/player/service/MediaScannerServiceTestCase.java]: Fixed

File: libresonic-main/src/test/java/org/libresonic/player/TestCaseUtils.java (hunk 6 above)
Comment (biconou) [inline libresonic-main/src/test/java/org/libresonic/player/TestCaseUtils.java]: It is necessary to wait until the scan process has finished.

File: libresonic-main/src/test/java/org/libresonic/player/service/MediaScannerServiceTestCase.java (hunk 7 above)
Comment (biconou) [inline libresonic-main/src/test/java/org/libresonic/player/service/MediaScannerServiceTestCase.java]: I added an assertion.

**Comment by biconou** [comment]
Fixed

**Comment by muff1nman** [comment]
Looks good, can you rebase and signoff?

**Comment by biconou** [comment]
@muff1nman by rebase, do you mean merge all commits in one unique ? 
I never used the signature, I'll learn. Any advice ?

**Comment by muff1nman** [comment]
I didn't mean the gpg signature which you see on my commits (-S flag for git). Definitely welcome to do that of course, but the main thing required by project guidelines is to signoff (--signoff) to acknowledge the project license. The two are mutually exclusive.

For the rebase I just mean rebase against the most recent develop branch, not squash into one commit. Usually I do this for people submitting PR's in the interest of time, but it makes it easier on me and youll need to edit your commits anyways for the signoff.

tl;dr I would do something like this.
`git rebase -i develop`, then change "pick" to "edit". For each commit, `git --amend --signoff --no-edit`

**Comment by biconou** [comment]
Rebased and signed-off
